server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java

```diff
@@ -178,47 +185,80 @@ public void clear() throws InterruptedException
   }
 
   /**
-   * Add a row. Must not be called concurrently from multiple threads.
+   * Add a row.  This method may internally incur persisting data added so far.  Also, if too large data are persisted,
```

Comment:
@gianm @pjain1 @dclim sounds good to me, but i have a question.

> Change ```finish``` to ```publish``` and have it return a ```Future<SegmentsAndMetadata>``` instead of ```SegmentsAndMetadata```

here, which operation results in ```Future<SegmentsAndMetadata>```? More precisely, you mean the new ```publish()``` should return a future for segment handoff? or pushing segments? Currently, ```push()``` is an async operation which returns a future for pushing segments, while ```publish()``` is an atomic sync operation which writes metadata to the underlying dbms.